File: web/packages/shared/src/types/skill.ts
export interface SkillTranslation {
  name: string;
  description: string;
}

export type SkillKnowledgeFileSourceType = "text" | "pdf_text";

export type SkillKnowledgeFileStatus =
  | "extracting"
  | "uploading"
  | "indexing"
  | "ready"
  | "failed"
  | "replacing"
  | "deleting"
  | "disabled";

export type SkillKnowledgeIngestionMode = "native_file" | "extracted_text";

export type SkillKnowledgeExtractedFrom = "xlsx";

export type SkillKnowledgeErrorCode =
  | "openai_not_configured"
  | "openai_endpoint_not_official"
  | "retrieval_model_not_enabled"
  | "knowledge_service_unavailable"
  | "unsupported_file_type"
  | "reference_file_too_large"
  | "knowledge_file_too_large"
  | "knowledge_total_size_exceeded"
  | "knowledge_extract_failed"
  | "knowledge_upload_failed"
  | "knowledge_index_failed"
  | "knowledge_retrieve_failed"
  | "knowledge_cleanup_failed";

export interface Skill {
  id: string;
  key?: string;
  name: string;
  description: string;
  translations?: Record<string, SkillTranslation>;
  icon: string;
  color: string;
  systemPrompt: string;
  suggestedProviderId?: string;
  suggestedModelId?: string;
  /** "any" | "reasoning" | "vision" | "fast" | "large-context" */
  modelCapabilityHint: string;
  temperature?: number;
  reasoningLevel?: string;
  webSearchEnabled?: boolean;
  starterMessages: string[];
  knowledgeFiles: SkillKnowledgeFile[];
  knowledgeBase: SkillKnowledgeBase | null;
  useMemory: boolean;
  isPinned: boolean;
  /** Pin order, updated on drag reorder */
  pinOrder: number;
  /** "builtin" | "user" | "community" */
  source: string;
  forkedFromId?: string;
  category?: string;
  /** Sort weight of a built-in skill within its category */
  sortOrder: number;
  usageCount: number;
  lastUsedAt?: string;
  createdAt: string;
  updatedAt: string;
}

export interface SkillKnowledgeFile {
  id: string;
  name: string;
  mimeType: string;
  sourceType: SkillKnowledgeFileSourceType;
  /** Plain text content; the client trims it to the context budget before sending */
  content: string;
  charCount: number;
  createdAt: string;
  updatedAt: string;
}

export interface SkillKnowledgeBase {
  provider: string;
  retrievalModel: string;
  vectorStoreId: string;
  expiresAfterDays: number;
  files: SkillKnowledgeBaseFile[];
  updatedAt: string;
}

export interface SkillKnowledgeBaseFile {
  id: string;
  name: string;
  mimeType: string;
  sizeBytes: number;
  ingestionMode: SkillKnowledgeIngestionMode;
  extractedFrom?: SkillKnowledgeExtractedFrom;
  openAIFileId?: string;
  status: SkillKnowledgeFileStatus;
  errorCode?: SkillKnowledgeErrorCode;
  createdAt: string;
  updatedAt: string;
}

export interface SkillUsage {
  count: number;
  /** null means no ceiling, which is what a local-only skill store always reports. */
  limit: number | null;
}

export interface SkillCategory {
  id: string;
  name: string;
  icon: string;
  sortOrder: number;
}

File: android/app/src/main/java/ai/oriveo/community/core/model/Skill.kt
package ai.oriveo.community.core.model

import androidx.compose.runtime.Immutable
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Immutable
@Serializable
data class Skill(
    val id: String,
    val key: String? = null,
    val name: String,
    val description: String = "",
    val icon: String = "\uD83E\uDD16",
    val color: String = "#6d38ff",
    val systemPrompt: String,
    val suggestedProviderId: String? = null,
    val suggestedModelId: String? = null,
    val modelCapabilityHint: String = "any",
    val temperature: Double? = null,
    val reasoningLevel: String? = null,
    val webSearchEnabled: Boolean? = null,
    val starterMessages: List<String> = emptyList(),
    val knowledgeFiles: List<SkillKnowledgeFile> = emptyList(),
    val knowledgeBase: SkillKnowledgeBase? = null,
    val useMemory: Boolean = true,
    val isPinned: Boolean = false,
    val pinOrder: Int = 0,
    val source: SkillSource = SkillSource.USER,
    val forkedFromId: String? = null,
    val category: String? = null,
    val sortOrder: Int = 0,
    val usageCount: Int = 0,
    val lastUsedAt: String? = null,
    val createdAt: String = "",
    val updatedAt: String = "",
) {
    val isBuiltIn get() = source == SkillSource.BUILTIN
    val isEditable get() = source == SkillSource.USER
}

@Immutable
@Serializable
data class SkillKnowledgeFile(
    val id: String,
    val name: String,
    val mimeType: String = "text/plain",
    val sourceType: SkillKnowledgeSourceType = SkillKnowledgeSourceType.TEXT,
    val content: String,
    val charCount: Int,
    val createdAt: String? = null,
    val updatedAt: String? = null,
)

@Serializable
enum class SkillKnowledgeSourceType {
    @SerialName("text") TEXT,
    @SerialName("pdf_text") PDF_TEXT,
}

@Serializable
enum class SkillKnowledgeFileStatus {
    @SerialName("extracting") EXTRACTING,
    @SerialName("uploading") UPLOADING,
    @SerialName("indexing") INDEXING,
    @SerialName("ready") READY,
    @SerialName("failed") FAILED,
    @SerialName("replacing") REPLACING,
    @SerialName("deleting") DELETING,
    @SerialName("disabled") DISABLED,
}

@Serializable
enum class SkillKnowledgeIngestionMode {
    @SerialName("native_file") NATIVE_FILE,
    @SerialName("extracted_text") EXTRACTED_TEXT,
}

@Serializable
enum class SkillKnowledgeExtractedFrom {
    @SerialName("xlsx") XLSX,
}

@Serializable
enum class SkillKnowledgeErrorCode {
    @SerialName("openai_not_configured") OPENAI_NOT_CONFIGURED,
    @SerialName("openai_endpoint_not_official") OPENAI_ENDPOINT_NOT_OFFICIAL,
    @SerialName("retrieval_model_not_enabled") RETRIEVAL_MODEL_NOT_ENABLED,
    @SerialName("knowledge_service_unavailable") KNOWLEDGE_SERVICE_UNAVAILABLE,
    @SerialName("unsupported_file_type") UNSUPPORTED_FILE_TYPE,
    @SerialName("reference_file_too_large") REFERENCE_FILE_TOO_LARGE,
    @SerialName("reference_file_char_limit_exceeded") REFERENCE_FILE_CHAR_LIMIT_EXCEEDED,
    @SerialName("knowledge_file_too_large") KNOWLEDGE_FILE_TOO_LARGE,
    @SerialName("knowledge_total_size_exceeded") KNOWLEDGE_TOTAL_SIZE_EXCEEDED,
    @SerialName("knowledge_extract_failed") KNOWLEDGE_EXTRACT_FAILED,
    @SerialName("knowledge_upload_failed") KNOWLEDGE_UPLOAD_FAILED,
    @SerialName("knowledge_index_failed") KNOWLEDGE_INDEX_FAILED,
    @SerialName("knowledge_retrieve_failed") KNOWLEDGE_RETRIEVE_FAILED,
    @SerialName("knowledge_cleanup_failed") KNOWLEDGE_CLEANUP_FAILED,
}

@Immutable
@Serializable
data class SkillKnowledgeBase(
    val provider: String,
    val retrievalModel: String,
    val vectorStoreId: String,
    val expiresAfterDays: Int,
    val files: List<SkillKnowledgeBaseFile> = emptyList(),
    val updatedAt: String? = null,
)

@Immutable
@Serializable
data class SkillKnowledgeBaseFile(
    val id: String,
    val name: String,
    val mimeType: String,
    val sizeBytes: Long,
    val ingestionMode: SkillKnowledgeIngestionMode,
    val extractedFrom: SkillKnowledgeExtractedFrom? = null,
    val openAIFileId: String? = null,
    val status: SkillKnowledgeFileStatus,
    val errorCode: SkillKnowledgeErrorCode? = null,
    val createdAt: String? = null,
    val updatedAt: String? = null,
)

@Serializable
enum class SkillSource {
    @SerialName("builtin") BUILTIN,
    @SerialName("user") USER,
    @SerialName("community") COMMUNITY;

    val value: String get() = name.lowercase()

    companion object {
        fun from(s: String) = entries.firstOrNull { it.value == s } ?: USER
    }
}

@Immutable
@Serializable
data class SkillCategory(
    val id: String,
    val name: String,
    val icon: String = "",
    val sortOrder: Int = 0,
)
